### PR TITLE
Support C++ in app-layer

### PR DESCRIPTION
**Pull request by jonasdn**
- Kbuild: Add support for C++ (.cc and .cpp) source
- app-layer: Add C++ example app

Fixes: #960

### CODE REVIEW

File: scripts/Makefile.lib [hunk 1]
```diff
@@ -70,8 +70,10 @@ _c_flags       = $(filter-out $(CFLAGS_REMOVE_$(basetarget).o), $(orig_c_flags))
 orig_a_flags   = $(KBUILD_CPPFLAGS) $(KBUILD_AFLAGS) $(KBUILD_SUBDIR_ASFLAGS) \
                  $(asflags-y) $(AFLAGS_$(basetarget).o)
 _a_flags       = $(filter-out $(AFLAGS_REMOVE_$(basetarget).o), $(orig_a_flags))
-_cpp_flags     = $(KBUILD_CPPFLAGS) $(cppflags-y) $(CPPFLAGS_$(@F))
+_cpp_flags     = $(KBUILD_CPPFLAGS)  $(KBUILD_CFLAGS) $(cppflags-y) $(CPPFLAGS_$(@F))
```
Comment (ntamas) [inline scripts/Makefile.lib:73]: Isn't `_cpp_flags` supposed to be the C _preprocessor_ flags only? Usually it is `CXXFLAGS` and similar that contains C++-specific flags.

Comment (jonasdn) [reply scripts/Makefile.lib:73]: You are right, I will make it more clear.

File: scripts/Makefile.build [hunk 2]
```diff
@@ -135,6 +135,9 @@ quiet_cmd_cc_o_c = CC $(quiet_modtag)  $@
 
 cmd_cc_o_c = $(CC) $(c_flags) -c -o $@ $<
 
+cmd_cxx_o_cpp = $(CXX)  $(cpp_flags) -c -o $@ $<
```
Comment (ntamas) [inline scripts/Makefile.build:138]: I needed to use `$(c_flags)` here to make things work, not `$(cpp_flags)`. In particular, include directories added in `Kbuild` files with `ccflags-y += ...` did not work until I changed this to `$(c_flags)`.

File: scripts/Makefile.lib [hunk 3]
```diff
@@ -70,8 +70,10 @@ _c_flags       = $(filter-out $(CFLAGS_REMOVE_$(basetarget).o), $(orig_c_flags))
...
 
+# Remove all non c++ compatible flags
+_cpp_flags     := $(filter-out -Wno-pointer-sign -Werror=implicit-int -std=gnu11,$(_cpp_flags))
```
Comment (ntamas) [inline scripts/Makefile.lib:76]: I needed to filter stuff from `_c_flags` and not `_cpp_flags` due to the other change I mentioned above.

**Comment by ntamas** [comment]
It works! At least with the changes I outlined above. The part that was not working is that I was adding an extra include folder to the build in `Kbuild` with `subdir-ccflags-y += -I$(srctree)/app/vendor/some-library/include` and it was not picked up until I switched to `_c_flags`.

Thanks for the quick fix!

File: scripts/Makefile.build (hunk 2 above)
Comment (jonasdn) [reply scripts/Makefile.build]: Could you try with latest added commit in PR?

File: scripts/Makefile.lib (hunk 1 above)
Comment (jonasdn) [reply scripts/Makefile.lib:73]: Fixed in later commit

**Comment by jonasdn** [comment]
> 

Thank you for testing it out! Could you try with the last commit: "Kbuild: Use cxx_flags for C++" ? To see if it solves your last issue.

File: scripts/Makefile.build [hunk 4]
```diff
@@ -181,6 +190,15 @@ $(obj)/%.o: $(src)/%.c $(objtool_obj) FORCE
 	$(call cmd,force_checksrc)
 	$(call if_changed_rule,cc_o_c)
 
+# Built-in and composite module parts
```
Comment (ntamas) [inline scripts/Makefile.build:193]: I think this line is not needed here; it was probably copied accidentally.

Comment (jonasdn) [reply scripts/Makefile.build]: You are correct. :)

File: scripts/Makefile.lib (hunk 1 above)
Comment (ntamas) [reply scripts/Makefile.lib:73]: Note that you should not add `$(KBUILD_CFLAGS)` then to `_cpp_flags` as `$(KBUILD_CFLAGS)` are flags for the C compiler and `_cpp_flags` are flags for the preprocessor.

**Comment by ntamas** [comment]
Yes, it works now! One tiny nitpick is that the output now shows `CC something.o` for C compilations but `arm-none-eabi-g++ something.o` for C++ files (and for some reason it includes the full path), but I couldn't figure out where this output is coming from. It doesn't bother me, but if you are super pedantic, this is the only issue left :)

**Comment by jonasdn** [comment]
> Yes, it works now! One tiny nitpick is that the output now shows `CC something.o` for C compilations but `arm-none-eabi-g++ something.o` for C++ files (and for some reason it includes the full path), but I couldn't figure out where this output is coming from. It doesn't bother me, but if you are super pedantic, this is the only issue left :)

Thanks!

It was a typo $(CXX) vs CXX, fixed in latest commit.

```
 CXX     /home/jonasdn/sandbox/kbuild-firmware/examples/app_hello_world-cpp/src/hello_world.o
```

**Review by tobbeanton — APPROVED** [review]
Looks good!

**Comment by jonasdn** [comment]
> > Yes, it works now! One tiny nitpick is that the output now shows `CC something.o` for C compilations but `arm-none-eabi-g++ something.o` for C++ files (and for some reason it includes the full path), but I couldn't figure out where this output is coming from. It doesn't bother me, but if you are super pedantic, this is the only issue left :)
> 
> Thanks!
> 
> It was a typo $(CXX) vs CXX, fixed in latest commit.
> 
> ```
>  CXX     /home/jonasdn/sandbox/kbuild-firmware/examples/app_hello_world-cpp/src/hello_world.o
> ```

The full-path I think is an artifact of the oot build.

**Comment by ntamas** [comment]
Before you guys merge this, here's a proposed patch to the build system that gets rid of some annoying `clang` warnings when building `Kbuild` itself on macOS:

```patch
diff --git a/tools/kbuild/Makefile.kbuild b/tools/kbuild/Makefile.kbuild
index 017ad491..fd7586c6 100644
--- a/tools/kbuild/Makefile.kbuild
+++ b/tools/kbuild/Makefile.kbuild
@@ -217,12 +217,17 @@ CONFIG_SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
 HOSTCC       = gcc
 HOSTCXX      = g++
 HOSTCFLAGS   = -Wall -Wmissing-prototypes -Wstrict-prototypes -O2
-HOSTCFLAGS  += -fomit-frame-pointer -Wno-format-overflow -std=gnu89
+HOSTCFLAGS  += -fomit-frame-pointer -std=gnu89
 HOSTCXXFLAGS = -O2

 ifeq ($(shell $(HOSTCC) -v 2>&1 | grep -c "clang version"), 1)
+# These flags are clang-only
 HOSTCFLAGS  += -Wno-unused-value -Wno-unused-parameter \
-               -Wno-missing-field-initializers -fno-delete-null-pointer-checks
+               -Wno-missing-field-initializers -fno-delete-null-pointer-checks \
+               -Wno-format-security
+else
+# These flags are gcc-only
+HOSTCFLAGS  += -Wno-format-overflow
 endif

 # Decide whether to build built-in, modular, or both.
```

(Sent this earlier on Discord).

**Comment by jonasdn** [comment]
> ccflags

Thanks I will test this a bit and add as a separate PR!

**Comment by jonasdn** [comment]
> Before you guys merge this, here's a proposed patch to the build system that gets rid of some annoying `clang` warnings when building `Kbuild` itself on macOS:
> 
> ```diff
> diff --git a/tools/kbuild/Makefile.kbuild b/tools/kbuild/Makefile.kbuild
> index 017ad491..fd7586c6 100644
> --- a/tools/kbuild/Makefile.kbuild
> +++ b/tools/kbuild/Makefile.kbuild
> @@ -217,12 +217,17 @@ CONFIG_SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
>  HOSTCC       = gcc
>  HOSTCXX      = g++
>  HOSTCFLAGS   = -Wall -Wmissing-prototypes -Wstrict-prototypes -O2
> -HOSTCFLAGS  += -fomit-frame-pointer -Wno-format-overflow -std=gnu89
> +HOSTCFLAGS  += -fomit-frame-pointer -std=gnu89
>  HOSTCXXFLAGS = -O2
> 
>  ifeq ($(shell $(HOSTCC) -v 2>&1 | grep -c "clang version"), 1)
> +# These flags are clang-only
>  HOSTCFLAGS  += -Wno-unused-value -Wno-unused-parameter \
> -               -Wno-missing-field-initializers -fno-delete-null-pointer-checks
> +               -Wno-missing-field-initializers -fno-delete-null-pointer-checks \
> +               -Wno-format-security
> +else
> +# These flags are gcc-only
> +HOSTCFLAGS  += -Wno-format-overflow
>  endif
> 
>  # Decide whether to build built-in, modular, or both.
> ```
> 
> (Sent this earlier on Discord).
> 
> _Edit_: this is on an Apple M1 CPU with clang 13, not sure whether it appears on Intel Macs.

Thanks I will test this a bit and add as a separate PR!

**Pull request merged**